finat/ufl/finiteelementbase.py

```diff
@@ -84,7 +85,7 @@ def _ufl_signature_data_(self):
 
     def __hash__(self):
         """Compute hash code for insertion in hashmaps."""
-        return hash(self._ufl_hash_data_())
+        return int.from_bytes(md5(self._ufl_hash_data_().encode()).digest())
```

Comment:
TODO: check if the different "variant"s hash differently

Reply:
dham made the same suggestion and I will double check/write a test. I think it should be fine since:

1. `self._ufl_hash_data_()` in `finiteelementbase.py` is
```python
def _ufl_hash_data_(self):
        """Doc."""
        return repr(self)
```

2. `repr(self)` of `FiniteElement` in `finiteelement.py` is
```python
def __repr__(self):
        """Format as string for evaluation as Python object."""
        return self._repr
```

3. And, finally, in `__init__` of `FiniteElement` in `finiteelement.py`
```python
self._repr = "FiniteElement(%s, %s, %s%s%s)" % (
        repr(self.family()), repr(self.cell), repr(self.degree()), quad_str, var_str)
```
where `var_str` contains the variant.

I'll come back to this once I've actually finished fixing the parallel caching behaviour in [PyOP2](https://github.com/OP2/PyOP2/pull/724)/[Firedrake](https://github.com/firedrakeproject/firedrake/pull/3730).

Reply:
Test added. I've only added a test for a couple of elements, but this should be sufficient to check the failure mode we saw previously.

I think this solution is sufficient when the finite element (Python object) is a subclass instance of `finat.ufl.finiteelementbase.FiniteElementBase` as these objects have a good repr. However, it isn't sufficient if the object is a subclass instance of `finat.finiteelementbase.FiniteElementBase` since the repr isn't implemented, contains a memory address and is very unsafe!

I'm trying to think of an instance where we might need to create a cache key based on the latter, otherwise I think this is sufficient. (@connorjward @dham)

Reply:
I think that until we meet a case where we need safe hashing for FInAT elements we should not spend time fixing them. Sometime in the near future @dham is going to do a refactor of this bit of the code so making changes now might be unproductive anyway.